phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -1189,40 +1237,168 @@ private Delete generateDeleteMarkers(List<Cell> row) {
+            if (indexPut == null) {
+                // This means the index row does not have any covered columns. We just need to check if the index row
+                // has only one cell (which is the empty column cell)
+                if (indexRow.rawCells().length == 1) {
+                    return;
+                }
+                exceptionMessage = "Index verify failed - Expected to find only empty column cell but got "
+                        + indexRow.rawCells().length + " cells - " + indexHTable.getName();
+                throw new IOException(exceptionMessage);
+            }
+            else {
+                // Remove the empty column prepared by Index codec as we need to change its value
+                removeEmptyColumn(indexPut, indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                        indexMaintainer.getEmptyKeyValueQualifier());
+            }
+            // Add the empty column
+            indexPut.addColumn(indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                    indexMaintainer.getEmptyKeyValueQualifier(), ts, VERIFIED_BYTES);
+            int cellCount = 0;
+            for (List<Cell> cells : indexPut.getFamilyCellMap().values()) {
+                if (cells == null) {
+                    break;
+                }
+                for (Cell expectedCell : cells) {
+                    Cell actualCell = indexRow.getColumnLatestCell(CellUtil.cloneFamily(expectedCell),
+                            CellUtil.cloneQualifier(expectedCell));
+                    if (actualCell == null) {
+                        exceptionMessage = "Index verify failed - Missing cell " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    // Check all columns
+                    if (!CellUtil.matchingValue(actualCell, expectedCell)) {
+                        exceptionMessage = "Index verify failed - Not matching cell value - " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    cellCount++;
+                }
+            }
+            if (cellCount != indexRow.rawCells().length) {
+                exceptionMessage = "Index verify failed - Expected to find " + cellCount + " cells but got "
+                        + indexRow.rawCells().length + " cells - " + indexHTable.getName();
+                throw new IOException(exceptionMessage);
+            }
+        }
+
+        private void verifyIndexRows(ArrayList<KeyRange> keys) throws IOException {
+            int expectedRowCount = keys.size();
+            ScanRanges scanRanges = ScanRanges.createPointLookup(keys);
+            Scan indexScan = new Scan();
+            indexScan.setTimeRange(scan.getTimeRange().getMin(), scan.getTimeRange().getMax());
```

Comment:
What is the index scan time-range? The max should be index rebuild start time (or whatever timestamp index is getting built at), right?

Reply:
It is the start time.